waku/v2/protocol/waku_noise/noise.nim

```diff
@@ -0,0 +1,168 @@
+## https://github.com/vacp2p/rfc/tree/master/content/docs/rfcs/35
+##
+## Implementation partially inspired by noise-libp2p:
+## https://github.com/status-im/nim-libp2p/blob/master/libp2p/protocols/secure/noise.nim
+
+{.push raises: [Defect].}
+
+import std/[oids, options, tables]
+import chronos
+import chronicles
+import bearssl
+import strutils
+import stew/[endians2]
+import nimcrypto/[utils, sha2, hmac]
+
+import libp2p/stream/[connection]
+import libp2p/peerid
+import libp2p/peerinfo
+import libp2p/protobuf/minprotobuf
+import libp2p/utility
+import libp2p/errors
+import libp2p/crypto/[crypto, chacha20poly1305, curve25519]
+
+
+when defined(libp2p_dump):
+  import libp2p/debugutils
+
+logScope:
+  topics = "nim-waku noise"
+
+const
+  # Empty is a special value which indicates k has not yet been initialized.
+  EmptyKey = default(ChaChaPolyKey)
+  NonceMax = uint64.high - 1 # max is reserved
+  NoiseSize = 32
+  MaxPlainSize = int(uint16.high - NoiseSize - ChaChaPolyTag.len)
+
+
+type
+  KeyPair* = object
+    privateKey: Curve25519Key
+    publicKey: Curve25519Key
+
+  NoisePublicKey* = object
+    flag: uint8
+    pk: seq[byte]
+
+  ChaChaPolyCiphertext* = object
+    data: seq[byte]
+    tag: ChaChaPolyTag
+
+  ChaChaPolyCipherState* = object
+    k*: ChaChaPolyKey
+    nonce*: ChaChaPolyNonce
+    ad*: seq[byte]
+
+  NoiseError* = object of LPError
+  NoiseHandshakeError* = object of NoiseError
+  NoiseDecryptTagError* = object of NoiseError
+  NoiseNonceMaxError* = object of NoiseError # drop connection on purpose
+  NoisePublicKeyError* = object of NoiseError
+  NoiseMalformedHandshake* = object of NoiseError
+
+
+
+#################################################################
+
+
+# ChaChaPoly encryption
+proc encrypt*(
+    state: ChaChaPolyCipherState,
+    plaintext: openArray[byte]): ChaChaPolyCiphertext
+    {.noinit, raises: [Defect].} =
+  #TODO: add padding
+  result.data.add plaintext
+  ChaChaPoly.encrypt(state.k, state.nonce, result.tag, result.data, state.ad)
+
+proc decrypt*(
+    state: ChaChaPolyCipherState, 
+    ciphertext: ChaChaPolyCiphertext): seq[byte]
+    {.raises: [Defect, NoiseDecryptTagError].} =
+  var
+    tagIn = ciphertext.tag
+    tagOut: ChaChaPolyTag
+  result = ciphertext.data
+  ChaChaPoly.decrypt(state.k, state.nonce, tagOut, result, state.ad)
+  #TODO: add unpadding
+  trace "decrypt", tagIn = tagIn.shortLog, tagOut = tagOut.shortLog, nonce = state.nonce
+  if tagIn != tagOut:
+    debug "decrypt failed", result = shortLog(result)
+    raise newException(NoiseDecryptTagError, "decrypt tag authentication failed.")
+
+
+proc randomChaChaPolyCipherState*(rng: var BrHmacDrbgContext): ChaChaPolyCipherState =
+  brHmacDrbgGenerate(rng, result.k)
+  brHmacDrbgGenerate(rng, result.nonce)
+  result.ad = newSeq[byte](32)
+  brHmacDrbgGenerate(rng, result.ad)
+
+
+#################################################################
+
+
+# Utility
+
+proc genKeyPair*(rng: var BrHmacDrbgContext): KeyPair =
```

Comment:
What is this key-pair used for? It is a DH key-pair?

Reply:
No specific use at this point, since we're just implementing some public keys logic. Added small proc comment https://github.com/status-im/nim-waku/commit/704f3fe40fb4ad63212ec5f6face687b4971c909